step/step.go

```diff
@@ -0,0 +1,566 @@
+				simulatorDevice = "iPad Air (3rd generation)"
+			}
+
+			sim, osVersion, errGetSimulator = s.simulatorManager.GetLatestSimulatorInfoAndVersion(platform, simulatorDevice)
+		} else {
+			normalizedOsVersion := input.SimulatorOsVersion
+			osVersionSplit := strings.Split(normalizedOsVersion, ".")
+			if len(osVersionSplit) > 2 {
+				normalizedOsVersion = strings.Join(osVersionSplit[0:2], ".")
+			}
+			osVersion = fmt.Sprintf("%s %s", platform, normalizedOsVersion)
+
+			sim, errGetSimulator = s.simulatorManager.GetSimulatorInfo(osVersion, input.SimulatorDevice)
+		}
+
+		if errGetSimulator != nil {
+			s.logger.Warnf("attempt %d to get simulator udid failed with error: %s", attempt, errGetSimulator)
+		}
+
+		return errGetSimulator
+	}); err != nil {
+		return simulator.Info{}, fmt.Errorf("simulator UDID lookup failed: %s", err)
+	}
+
+	s.logger.Infof("Simulator infos")
+	s.logger.Printf("* simulator_name: %s, version: %s, UDID: %s, status: %s", sim.Name, osVersion, sim.ID, sim.Status)
+
+	return sim, nil
+}
+
+// ProcessConfig ...
+func (s XcodeTestRunner) ProcessConfig() (Config, error) {
+	var input Input
+	err := s.inputParser.Parse(&input)
+	if err != nil {
+		return Config{}, err
+	}
+
+	stepconf.Print(input)
+	s.logger.Println()
+
+	s.logger.EnableDebugLog(input.Verbose)
+
+	// validate Xcode version
+	xcodebuildVersion, err := s.xcodebuild.Version()
+	if err != nil {
+		return Config{}, fmt.Errorf("failed to determine Xcode version, error: %s", err)
+	}
+	s.logger.Printf("- xcodebuildVersion: %s (%s)", xcodebuildVersion.Version, xcodebuildVersion.BuildVersion)
+
+	if err := s.validateXcodeVersion(&input, int(xcodebuildVersion.MajorVersion)); err != nil {
+		return Config{}, err
+	}
+
+	// validate project path
+	projectPath, err := s.pathModifier.AbsPath(input.ProjectPath)
+	if err != nil {
+		return Config{}, fmt.Errorf("failed to get absolute project path, error: %s", err)
+	}
+	if filepath.Ext(projectPath) != ".xcodeproj" && filepath.Ext(projectPath) != ".xcworkspace" {
+		return Config{}, fmt.Errorf("invalid project file (%s), extension should be (.xcodeproj/.xcworkspace)", projectPath)
+	}
+
+	// validate simulator related inputs
+	sim, err := s.validateSimulator(input)
+	if err != nil {
+		return Config{}, err
+	}
+
+	// validate test repetition related inputs
+	if input.TestRepetitionMode != xcodebuild.TestRepetitionNone && input.MaximumTestRepetitions < 2 {
+		return Config{}, fmt.Errorf("invalid number of Maximum Test Repetitions (maximum_test_repetitions): %d, should be more than 1", input.MaximumTestRepetitions)
+	}
+
+	if input.RelaunchTestsForEachRepetition && input.TestRepetitionMode == xcodebuild.TestRepetitionNone {
+		return Config{}, errors.New("Relaunch Tests for Each Repetition (relaunch_tests_for_each_repetition) cannot be used if Test Repetition Mode (test_repetition_mode) is 'none'")
+	}
+
+	return Config{
```

Comment:
I'd probably extract this to hide at the bottom and improve readability

Reply:
And generally in other cases
![image](https://user-images.githubusercontent.com/13942888/132199341-b8aecc71-cac4-46fd-9efd-1d276b46ff07.png)
![image](https://user-images.githubusercontent.com/13942888/132199346-24031c8b-aa37-4709-a8ea-febd57f6747d.png)

Reply:
addressed in https://github.com/bitrise-steplib/steps-xcode-test/pull/181/commits/17d18dfb94ac4fa2a882d071c890a4b700366ac6 and https://github.com/bitrise-steplib/steps-xcode-test/pull/181/commits/6774bfedae99181e8b1bee0ade4274f1f457f588